man/irccd.1
.\"
.\" Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd @IRCCD_MAN_DATE@
.Dt IRCCD 1
.Os
.\" NAME
.Sh NAME
.Nm irccd
.Nd IRC Client Daemon
.\" SYNOPSIS
.Sh SYNOPSIS
.Nm
.Op Fl v
.Op Fl c Ar file
.Nm
.Ar info
.Nm
.Ar paths
.Nm
.Ar version
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
program is an IRC bot which connects to one or more severs and
dispatches events to plugins and connected clients.
.Pp
The following options are available:
.Bl -tag -width indent-xxx
.It Fl c Ar file
specify the configuration file.
.It Fl v
be verbose.
.El
.Pp
When ran without arguments,
.Nm
will read your configuration file and dispatch IRC events to the plugins and
connected clients indefinitely.
.Pp
Otherwise, the following commands are available:
.Bl -tag -width 12n
.It Cm info
Show in a scriptable manner the options that were selected during irccd's build.
.It Cm version
Get the irccd version. if Mercurial was available when building irccd then the
current revision is bundled in the output.
.It Cm paths
Shows where irccd will load plugins, configuration files and such.
.El
.\" PLUGINS
.Sh PLUGINS
The
.Nm
program can run plugins once IRC events are received. For example, if someone
sends you a private message plugins will be invoked with that event. Both
native plugins written in C or C++ and Javascript are supported (if enabled at
compile time).
.Pp
The following IRC events are supported:
.Bl -tag -width 12n
.\" onCommand
.It onCommand
This is a special event that does not exist in IRC context. It calls the plugin
special invocation command using
.Ar "!name arguments..."
syntax. The exclamation mark is configured by default to be the command
character prefix, then if the adjacent name is known to be a loaded plugin it is
invoked with the additional arguments.
.Pp
Example: to call the
.Nm hangman
plugin, one may use !hangman to start a game.
.Pp
See also
.Xr irccd.conf 5
on how to change the command character prefix under a
.Va server
section block.
.\" onConnect
.It onConnect
When a server successfully connects to an IRC server.
.\" onDisconnect
.It onDisconnect
When a server disconnected from an IRC server both in case of failures or
explicit user disconnection.
.\" onInvite
.It onInvite
Event called when the bot itself has been invited to a channel.
.\" onJoin
.It onJoin
When someone joins a channel.
.\" onKick
.It onKick
When someone has been kicked from a channel, irccd may be included.
.\" onLoad
.It onLoad
This is a special event that does not exist in IRC context. It is invoked when
the plugin is initialized.
.\" onMessage
.It onMessage
Upon private message.
.\" onMe
.It onMe
On action emote, also most known as
.Ar /me
command.
.\" onMode
.It onMode
When a user or channel mode change.
.\" onNames
.It onNames
When a list of nicknames has been received.
.\" onNick
.It onNick
On nick change, irccd may be included.
.\" onNotice
.It onNotice
On private notice.
.\" onPart
.It onPart
When someone leaves a channel.
.\" onReload
.It onReload
This is a special event that does not exist in IRC context. It is invoked when
the user asks to reload a plugin.
.\" onTopic
.It onTopic
When a channel topic has been changed.
.\" onUnload
.It onUnload
This is a special event that does not exist in IRC context. It is invoked when
the user asks to unload a plugin and before exiting.
.\" onWhois
.It onWhois
When a whois information has been received.
.El
.Pp
The following plugins are provided with irccd:
.Pp
.Bl -bullet -compact
.It
ask
.It
auth
.It
hangman
.It
history
.It
joke
.It
links
.It
logger
.It
plugin
.It
roulette
.It
tictactoe
.El
.Pp
See additional documentation in their own manual page in the form
.Xr irccd-plugin-name 7
where name is the actual plugin name.
.\" HOOKS
.Sh HOOKS
Hooks are a different and more lightweight approach to plugins, they are
executed upon incoming events and spawned each time a new event arrives.
.Pp
In contrast to plugins, differences are:
.Pp
.Bl -bullet -compact
.It
Hooks can not be filtered with rules.
.It
Hooks does not support all events. These events are not supported:
.Em onLoad , onUnload , onReload , onCommand , onNames , onWhois .
.It
Hooks can be written in any language.
.It
Execution may be slower since scripting languages require to fire up the
interpreter each time a new event is available.
.El
.Pp
Each hook will receive as positional argument the event name (similar to plugin
events) and the event arguments.
.Pp
See also the section
.Va hooks
in
.Xr irccd.conf 5
manual page to enable hooks.
.\" TRANSPORTS
.Sh TRANSPORTS
The daemon can be controlled at runtime using the dedicated
.Nm irccdctl
tool or using raw TCP messages on a UNIX domain socket file.
.Pp
See also the
.Va transport
section in the
.Xr irccd.conf 5
manual page.
.\" RULES
.Sh RULES
.Nm
supports a feature called rules which allows you to define a fine-grained set of
rules allowed for specific plugins. For instance, you may want to disable some
IRC events for some plugins depending on your set of parameters. This is useful
for plugins that generates huge traffic.
.Pp
Rule events are matched using the same name as plugin events described in the
section above. For example, to disable a private message event you must use the
.Ar onMessage
value.
.Pp
See also the
.Va rule
section in the
.Xr irccd.conf 5
manual page.
.\" SEE ALSO
.Sh SEE ALSO
.Xr irccd-api 7 ,
.Xr irccd-cmake 7 ,
.Xr irccd-ipc 7 ,
.Xr irccd-plugin-ask 7 ,
.Xr irccd-plugin-auth 7 ,
.Xr irccd-plugin-hangman 7 ,
.Xr irccd-plugin-history 7 ,
.Xr irccd-plugin-joke 7 ,
.Xr irccd-plugin-links 7 ,
.Xr irccd-plugin-logger 7 ,
.Xr irccd-plugin-plugin 7 ,
.Xr irccd-plugin-roulette 7 ,
.Xr irccd-plugin-tictactoe 7 ,
.Xr irccd-templates 7 ,
.Xr irccd.conf 5 ,
.Xr irccdctl 1 ,
.Xr irccdctl.conf 5
.\" AUTHORS
.Sh AUTHORS
The
.Nm
daemon was written by
.An David Demelier Aq Mt markand@malikania.fr .
.\" AUTHORS
.Sh AUTHORS
The
.Nm irccd
daemon was written by
.An David Demelier Aq Mt markand@malikania.fr .


CHANGES.md
IRC Client Daemon CHANGES
=========================

irccd 5.0.0 ????-??-??
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight of this release is the integration of libev as the main loop
and coroutines, making the code simpler and more expressive.

Build system
------------

The build system has been changed from CMake to GNU Make, this is because CMake
makes simple things complicated.

C API
-----

- Most of the string based types are dynamically allocated rather than creating
  dozens of arbitrary fixed size limits.
- Functions for allocated types are now coupled as `foo_new` and `foo_free` or
  `foo_init` and `foo_finish` otherwise.
- Types that require string list are now implemented using NULL sentinel based
  lists.

irccd
-----

- The auto-reconnect feature is automatically enabled on a server and is no
  longer an option. As an IRC bot, it make sense that a server should be
  reconnected permanently.
- All deferred functions have been removed.
- The server connection procedure is now implemented using coroutines for a
  more readable code logic.

misc
----

The C and Javascript APIs are now documented using doxygen rather than manual
pages.

irccd 4.1.1 2026-01-07
======================

- Fix minor issue in Irccd.Http API when `curl_multi_fdset` can return `-1` file
  descriptor.

irccd 4.1.0 2025-04-23
======================

- Add `irc_bot_pollable_add` to insert a custom interface into the main irccd
  loop.
- A new option `IRCCD_WITH_HTTP` has been added to accommodate the new curl
  dependency for `Irccd.Http` API.
- The plugin `history` manages two different timestamps for message and
  visibility. This helps showing the last message date of someone who is present
  on a channel but idle.

javascript API
--------------

- Brand new asynchronous `Irccd.Http` API.

plugins
-------

- links: rewritten in Javascript using the brand new `Irccd.Http` API.

irccd 4.0.4 2023-08-03
======================

- Improve tests running on a slow machines.
- Fix a warning on a invalid comparison.

irccd 4.0.3 2023-02-08
======================

- Fix build on NetBSD 9.3.

irccd 4.0.2 2023-01-18
======================

- Fix crash because of undefined variable.

irccd 4.0.1 2022-02-07
======================

- Fix disconnection with multiple servers.
- Fix Javascript `Irccd.Server` constructor function.
- Substitute the irccd version in Javascript plugins.
- Fix `irccdctl plugin-reload` command without argument.

irccd 4.0.0 2022-02-03
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight for this release is the rewrite from C++ to C. The only
runtime dependency required is OpenSSL (if built with SSL support).

irccd
-----

- Irccd keeps track of nicknames in channels by capturing join/part/kick and
  mode changes. It is now more convenient from the plugins to quickly inspect if
  someone is present on a channel.
- It is now possible to change uid/gid of the transport socket file. The file is
  also created with permissions 664.
- A new `paths` command has been added to show default paths.

irccdctl
--------

- Commands `plugin-reload` and `plugin-unload` can be invoked without arguments.
- New `plugin-template` and `plugin-path` command which are synonyms of
  `plugin-config` but for templates and paths respectively.

plugins
-------

- tictactoe: now has a timeout in case of inactivity.

misc
----

- Split irccd-api manual page into individual irccd-api-<module> for a better
  readability.
- New `irccd.conf` and `irccdctl.conf` syntax.

network API
-----------

- Network protocol uses plain text again.
- Transport uses clear UNIX sockets only without passwords.

javascript API
--------------

- Brand new Irccd.Rule API to inspect and manage rules.
- Brand new Irccd.Hook API to inspect and manage hooks.

irccd 3.1.1 2021-01-04
======================

- Synchronize `ping-timeout` option in `[server]` to 1800 seconds by default.
- Enable `auto-reconnect` option in `[server]` by default as specified in the
  manual page.

irccd 3.1.0 2020-07-03
======================

- Added a new hook system. Hooks consist of an alternative approach to plugins
  to extend irccd in any language.

irccd 3.0.3 2019-10-06
======================

- Fix errors in irccdctl.conf example file.
- Add example of password in irccdctl.conf and irccd.conf.

irccd 3.0.2 2019-09-22
======================

- Added *IRCCD_WITH_JS* CMake variable in irccd package.
- Fixed trailing CTCP escape code.
- Fixed invalid parameters in topic event.
- Fixed invalid documentation of Irccd.System.usleep function.
- Fixed invalid system configuration directory.

irccd 3.0.1 2019-09-01
======================

- Fixed an invalid template escape sequence.
- Updated the default configuration files.
- Fix RPATH handling for private libraries like Duktape.

irccd 3.0.0 2019-08-15
======================

This is a major release. See MIGRATING.md file for more information.

irccd
-----

- New sections `[paths]` and `[paths.plugin]` have been added to control
  standard paths for both irccd and plugins.
- Irccd no longer supports uid, gid, pid and daemon features.
- Sections `[identity]` and `[server]` have been merged.
- Local transports support SSL.
- The origin in rule is now first class value.
- New option `ipv4` in `[transport]`.
- New option `ipv4` in `[server]`.
- Section `[format]` is renamed to `[templates]`.
- New commands are available as irccd arguments `info` and `version`.

irccdctl
--------

- New option `ipv4` in `[connect]`.
- New option `-o` in `rule-add`.
- New option `-o` and `-O` in `rule-edit`.

irccd-test
----------

- A brand new `irccd-test` program has been added to tests plugins on the
  command line.

cmake
-----

- CMake no longer create a fake installation directory while building.
- All targets are placed into the `bin` directory while building.

network API
-----------


lib/irccd/irccd.c
/*
 * irccd.c -- main irccd object
 *
 * Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <assert.h>
#include <ctype.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <utlist.h>

#include <ev.h>

#include "config.h"
#include "event.h"
#include "hook.h"
#include "irccd.h"
#include "log.h"
#include "plugin.h"
#include "rule.h"
#include "server.h"
#include "util.h"

/* Public bot context. */
static struct irccd bot = {};

const struct irccd *irccd = &bot;

static int
is_command(const struct irc_plugin *p, const struct irc_event *ev)
{
	const char *cc;
	size_t ccsz;

	if (ev->type != IRC_EVENT_MESSAGE)
		return 0;

	/* Get the command prefix (e.g !)*/
	cc = ev->server->prefix;
	ccsz = strlen(cc);

	return strncmp(ev->message.message, cc, ccsz) == 0 &&
	       strncmp(ev->message.message + ccsz, p->name, strlen(p->name)) == 0;
}

static struct irc_event *
to_command(const struct irc_plugin *p, const struct irc_event *ev)
{
	static struct irc_event cev;

	/* Convert "!test foo bar" to "foo bar" */
	memcpy(&cev, ev, sizeof (*ev));
	cev.type = IRC_EVENT_COMMAND;
	cev.message.message += strlen(cev.server->prefix) + strlen(p->name);

	while (*cev.message.message && isspace(*cev.message.message))
		++cev.message.message;

	return &cev;
}

static int
invokable(const struct irc_plugin *p, const struct irc_event *ev)
{
	switch (ev->type) {
	case IRC_EVENT_COMMAND:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->message.channel, ev->message.origin, p->name, "onCommand");
	case IRC_EVENT_CONNECT:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    NULL, NULL, p->name, "onConnect");
	case IRC_EVENT_DISCONNECT:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    NULL, NULL, p->name, "onDisconnect");
	case IRC_EVENT_INVITE:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->invite.channel, ev->invite.origin, p->name, "onInvite");
	case IRC_EVENT_JOIN:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->join.channel, ev->join.origin, p->name, "onJoin");
	case IRC_EVENT_KICK:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->kick.channel, ev->kick.origin, p->name, "onKick");
		break;
	case IRC_EVENT_ME:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->message.channel, ev->message.origin, p->name, "onMe");
	case IRC_EVENT_MESSAGE:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->message.channel, ev->message.origin, p->name, "onMessage");
	case IRC_EVENT_MODE:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->mode.channel, ev->mode.origin, p->name, "onMode");
	case IRC_EVENT_NAMES:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->names.channel, NULL, p->name, "onNames");
	case IRC_EVENT_NICK:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    NULL, ev->nick.origin, p->name, "onNick");
	case IRC_EVENT_NOTICE:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->notice.channel, ev->notice.origin, p->name, "onNotice");
	case IRC_EVENT_PART:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->part.channel, ev->part.origin, p->name, "onPart");
	case IRC_EVENT_TOPIC:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    ev->topic.channel, ev->topic.origin, p->name, "onTopic");
	case IRC_EVENT_WHOIS:
		return irc_rule_matchlist(bot.rules, ev->server->name,
		    NULL, NULL, p->name, "onWhois");
	default:
		return 1;
	}
}

static struct irc_plugin *
try_plugin(struct irc_plugin_loader *ldr, const char *base, const char *name, const char *ext)
{
	char path[PATH_MAX] = {};

	if (ext)
		snprintf(path, sizeof (path), "%s/%s.%s", base, name, ext);
	else
		snprintf(path, sizeof (path), "%s/%s", base, name);

	irc_log_info("irccd: trying %s", path);

	return irc_plugin_loader_open(ldr, name, path);
}

static struct irc_plugin *
find_plugin(struct irc_plugin_loader *ldr, const char *base, const char *name)
{
	char *extensions = NULL, *t, *ext;
	struct irc_plugin *p;

	/* Copy the extensions to iterate over ':' */
	if (ldr->extensions) {
		extensions = irc_util_strdup(ldr->extensions);

		for (t = extensions; (ext = strtok_r(t, ":", &t)); )
			if ((p = try_plugin(ldr, base, name, ext)))
				break;

		free(extensions);
	} else {
		/*
		 * No extension? weird but allow a unique direct filename in the
		 * directory mentioned.
		 */
		p = try_plugin(ldr, base, name, NULL);
	}

	return p;
}

static inline struct irc_plugin *
open_plugin(struct irc_plugin_loader *ldr, const char *name, const char *path)
{
	return irc_plugin_loader_open(ldr, name, path);
}

static inline int
is_extension_valid(const struct irc_plugin_loader *ldr, const char *path)
{
	char *extensions, *token, *p, *ext;
	int valid = 0;

	/* If we're unable to find an extension, assume it's allowed. */
	if (!(ext = strrchr(path, '.')))
		return 1;

	extensions = irc_util_strdup(ldr->extensions);
	ext++;

	for (p = extensions; (token = strtok_r(p, ":", &p)); ) {
		if (strcmp(token, ext) == 0) {
			valid = 1;
			break;
		}
	}

	free(extensions);

	return valid;
}

void
irc_bot_init(void)
{
	irc_log_to_console();
}

void *
irc_bot_loop(void)
{
	return NULL;
}

int
irc_bot_server_add(struct irc_server *s)
{
	assert(s);

	if (irc_bot_server_get(s->name)) {
		irc_log_warn("irccd: server %s already exists", s->name);
		return -1;
	}

	irc_log_info("irccd: added new server: %s", s->name);

	irc_server_incref(s);
	irc_server_connect(s);

	LL_APPEND(bot.servers, s);

	return 0;
}

struct irc_server *
irc_bot_server_get(const char *name)
{
	struct irc_server *s;

	DL_FOREACH(bot.servers, s)
		if (strcmp(s->name, name) == 0)
			return s;

	return NULL;
}

void
irc_bot_server_remove(const char *name)
{
	struct irc_server *s;

	if (!(s = irc_bot_server_get(name)))
		return;

	irc_server_disconnect(s);

	/* Don't forget to notify plugins. */
	irc_bot_dispatch(&(struct irc_event) {
		.type = IRC_EVENT_DISCONNECT,
		.server = s
	});

	LL_DELETE(bot.servers, s);
	irc_server_decref(s);
}

void
irc_bot_server_clear(void)
{
	struct irc_server *s, *tmp;

	LL_FOREACH_SAFE(bot.servers, s, tmp)
		irc_bot_server_remove(s->name);

	bot.servers = NULL;
}

int
irc_bot_plugin_add(struct irc_plugin *p)
{
	assert(p);

	int rc;

	if (irc_bot_plugin_get(p->name)) {
		irc_log_warn("irccd: plugin %s already exists", p->name);
		return -1;
	}

	if ((rc = irc_plugin_load(p)) == 0) {
		LL_PREPEND(bot.plugins, p);
		irc_log_info("irccd: add new plugin: %s (%s)", p->name, p->description);
		irc_log_info("irccd: %s: version %s, from %s (%s license)", p->name,
		    p->version, p->author, p->license);
	} else
		irc_log_warn("irccd: plugin %s failed to load", p->name);

	return rc;
}

struct irc_plugin *
irc_bot_plugin_search(const char *name, const char *path)
{
	assert(name);

	char *paths = NULL, pathbuf[PATH_MAX], *t, *token;
	struct irc_plugin *p = NULL;
	struct irc_plugin_loader *ldr;

	if (!path)
		irc_log_info("irccd: trying to find plugin %s", name);
	else
		irc_log_info("irccd: opening plugin %s", name);

	LL_FOREACH(bot.plugin_loaders, ldr) {
		if (p)
			break;

		if (path) {
			if (!is_extension_valid(ldr, path))
				continue;

			p = open_plugin(ldr, name, path);
		} else if (paths) {
			/* Copy the paths to tokenize it. */
			paths = irc_util_strdup(ldr->paths);

			/*
			 * For every directory (separated by colon) call find_plugin
			 * which will append the extension and try to open it.
			 */
			for (t = paths; (token = strtok_r(t, ":", &t)); ) {
				if ((p = find_plugin(ldr, token, name)))
					break;
			}
		}
	}

	free(paths);

	if (!p) {
		irc_log_warn("irccd: could not find plugin %s", name);
		return NULL;
	}

	/* Set default paths if they are not set. */
	irc_plugin_set_path(p, "cache", irc_util_printf(pathbuf, sizeof (pathbuf),
	    "%s/plugin/%s", IRCCD_CACHEDIR, p->name));
	irc_plugin_set_path(p, "data", irc_util_printf(pathbuf, sizeof (pathbuf),
	    "%s/plugin/%s", IRCCD_DATADIR, p->name));
	irc_plugin_set_path(p, "config", irc_util_printf(pathbuf, sizeof (pathbuf),
	    "%s/irccd/plugin/%s", IRCCD_SYSCONFDIR, p->name));

	return p;
}

struct irc_plugin *
irc_bot_plugin_get(const char *name)
{
	struct irc_plugin *p;

	LL_FOREACH(bot.plugins, p)
		if (strcmp(p->name, name) == 0)
			return p;

	return NULL;
}

void
irc_bot_plugin_remove(const char *name)
{
	struct irc_plugin *p;

	if (!(p = irc_bot_plugin_get(name)))
		return;

	LL_DELETE(bot.plugins, p);
	irc_plugin_unload(p);
	irc_plugin_finish(p);
}

void
irc_bot_plugin_loader_add(struct irc_plugin_loader *ldr)
{
	assert(ldr);

	LL_PREPEND(bot.plugin_loaders, ldr);
}

void
irc_bot_plugin_clear(void)
{
	struct irc_plugin *p, *tmp;

	LL_FOREACH_SAFE(bot.plugins, p, tmp)
		irc_bot_plugin_remove(p->name);

	bot.plugins = NULL;
}

void
irc_bot_rule_insert(struct irc_rule *rule, size_t index)
{
	assert(rule);

	if (index == 0)
		DL_PREPEND(bot.rules, rule);
	else if (index >= irc_bot_rule_size())
		DL_APPEND(bot.rules, rule);
	else {
		struct irc_rule *pos;

		for (pos = bot.rules; --index; )
			pos = pos->next;

		DL_APPEND_ELEM(bot.rules, pos, rule);
	}
}

struct irc_rule *
irc_bot_rule_get(size_t index)
{
	assert(index < irc_bot_rule_size());

	struct irc_rule *rule;

	for (rule = bot.rules; index-- != 0; )
		rule = rule->next;

	return rule;
}

void
irc_bot_rule_move(size_t from, size_t to)
{
	assert(from < irc_bot_rule_size());

	struct irc_rule *f, *t;

	if (from == to)
		return;

	f = t = bot.rules;

	while (from--)
		f = f->next;

	DL_DELETE(bot.rules, f);

	if (to == 0)
		DL_PREPEND(bot.rules, f);
	else {
		while (t && to--)
			t = t->next;

		DL_APPEND_ELEM(bot.rules, t, f);
	}
}

void
irc_bot_rule_remove(size_t index)
{
	assert(index < irc_bot_rule_size());

	struct irc_rule *pos = bot.rules;

	for (size_t i = 0; i < index; ++i)
		pos = pos->next;

	DL_DELETE(bot.rules, pos);
}

size_t
irc_bot_rule_size(void)
{
	const struct irc_rule *r;
	size_t total = 0;

	DL_FOREACH(bot.rules, r)
		total++;

	return total;
}

void
irc_bot_rule_clear(void)
{
	struct irc_rule *r, *tmp;

	DL_FOREACH_SAFE(bot.rules, r, tmp)
		irc_rule_free(r);

	bot.rules = NULL;
}

int
irc_bot_hook_add(struct irc_hook *h)
{
	assert(h);

	if (irc_bot_hook_get(h->name)) {
		irc_log_warn("irccd: hook %s already exists", h->name);
		return -1;
	}

	LL_PREPEND(bot.hooks, h);

	return 0;
}

struct irc_hook *
irc_bot_hook_get(const char *name)
{
	struct irc_hook *h;

	LL_FOREACH(bot.hooks, h)
		if (strcmp(h->name, name) == 0)
			return h;

	return NULL;
}

void
irc_bot_hook_remove(const char *name)
{
	assert(name);

	struct irc_hook *h;

	if ((h = irc_bot_hook_get(name))) {
		LL_DELETE(bot.hooks, h);
		irc_hook_free(h);
	}
}

void
irc_bot_hook_clear(void)
{
	struct irc_hook *h, *tmp;

	LL_FOREACH_SAFE(bot.hooks, h, tmp)
		irc_hook_free(h);

	bot.hooks = NULL;
}

void
irc_bot_dispatch(const struct irc_event *ev)
{
	struct irc_plugin *p, *ptmp, *plgcmd = NULL;
	struct irc_hook *h, *htmp;

	LL_FOREACH_SAFE(bot.hooks, h, htmp)
		irc_hook_invoke(h, ev);

	/*
	 * Invoke for every plugin the event verbatim. Then, the event may match
	 * a plugin name command in that case we need to modify the event but
	 * only one plugin can match by its identifier. For example, the
	 * following plugins are loaded:
	 *
	 * - ask
	 * - hangman
	 * - logger
	 *
	 * If the message is "!ask will I be reach?" then it will invoke
	 * onMessage for hangman and logger but onCommand for ask. As such call
	 * hangman and logger first and modify event before ask.
	 */
	LL_FOREACH_SAFE(bot.plugins, p, ptmp) {
		if (is_command(p, ev))
			plgcmd = p;
		else if (invokable(p, ev))
			irc_plugin_handle(p, ev);
	}

	if (plgcmd && invokable(plgcmd, ev))
		irc_plugin_handle(plgcmd, to_command(plgcmd, ev));
}

void
irc_bot_finish(void)
{
	struct irc_plugin_loader *ld, *ldtmp;

	/*
	 * First remove all loaders to make sure plugins won't try to load
	 * new plugins.
	 */
	LL_FOREACH_SAFE(bot.plugin_loaders, ld, ldtmp)
		irc_plugin_loader_finish(ld);

	bot.plugin_loaders = NULL;

	irc_bot_server_clear();
	irc_bot_plugin_clear();
	irc_bot_hook_clear();
	irc_bot_rule_clear();
}
